sdk/dotnet/cmd/pulumi-language-dotnet/main.go

```diff
@@ -117,37 +118,231 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
+	}
+
+	return &pulumirpc.GetRequiredPluginsResponse{Plugins: plugins}, nil
+}
+
+func (host *dotnetLanguageHost) DeterminePulumiPackages(ctx context.Context) ([][]string, error) {
+	logging.V(5).Infof("GetRequiredPlugins: Determining pulumi packages")
+
+	// Run the `dotnet list package --include-transitive` command.  Importantly, do not clutter the
+	// stream with the extra steps we're performing. This is just so we can determine the required
+	// plugins.  And, after the first time we do this, subsequent runs will see that the plugin is
+	// installed locally and not need to do anything.
+	args := []string{"list", "package", "--include-transitive"}
+	commandStr := strings.Join(args, " ")
+	commandOutput, err := host.RunDotnetCommand(ctx, args, false /*logToUser*/)
+	if err != nil {
+		return nil, err
+	}
+
+	// expected output should be like so:
+	//
+	//    Project 'Aliases' has the following package references
+	//    [netcoreapp3.0]:
+	//    Top-level Package      Requested                        Resolved
+	//    > Pulumi               1.5.0-preview-alpha.1572911568   1.5.0-preview-alpha.1572911568
+	//
+	//    Transitive Package                                       Resolved
+	//    > Google.Protobuf                                        3.10.0
+	//    > Grpc                                                   2.24.0
+	outputLines := strings.Split(strings.Replace(commandOutput, "\r\n", "\n", -1), "\n")
+
+	sawPulumi := false
+	pulumiPackages := [][]string{}
+	for _, line := range outputLines {
+		fields := strings.Fields(line)
+		if len(fields) < 3 {
+			continue
+		}
+
+		// Has to start with `>` and have at least 3 chunks:
+		//
+		//    > name requested_ver? resolved_ver
+		if fields[0] != ">" {
+			continue
+		}
+
+		// We only care about `Pulumi.` packages
+		packageName := fields[1]
+		if packageName == "Pulumi" {
+			sawPulumi = true
+			continue
+		}
+
+		if !strings.HasPrefix(packageName, "Pulumi.") {
+			continue
+		}
+
+		version := fields[len(fields)-1]
+		pulumiPackages = append(pulumiPackages, []string{packageName, version})
+	}
+
+	if !sawPulumi && len(pulumiPackages) == 0 {
+		return nil, errors.Errorf(
+			"Unexpected output from 'dotnet %v'. Program does not appear to reference any 'Pulumi.*' packages.",
+			commandStr)
+	}
+
+	logging.V(5).Infof("GetRequiredPlugins: Pulumi packages: %#v", pulumiPackages)
+
+	return pulumiPackages, nil
+}
+
+func (host *dotnetLanguageHost) DetermineDotnetPackageDirectory(ctx context.Context) (string, error) {
+	logging.V(5).Infof("GetRequiredPlugins: Determining package directory")
+
+	// Run the `dotnet nuget locals global-packages --list` command.  Importantly, do not clutter
+	// the stream with the extra steps we're performing. This is just so we can determine the
+	// required plugins.  And, after the first time we do this, subsequent runs will see that the
+	// plugin is installed locally and not need to do anything.
+	args := []string{"nuget", "locals", "global-packages", "--list"}
```

Comment:
Is there a chance that the user won't have packages installed globally?

Reply:
From https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders:

> The global-packages folder is where NuGet installs any downloaded package. Each package is fully expanded into a subfolder that matches the package identifier and version number. Projects using the PackageReference format **always use packages directly from this folder.**

Emphasis mine.

There are no other places that nuget should be looking from.

Reply:
Perfect; thanks!